### PR TITLE
Add default value to Mesh::get for ints

### PR DESCRIPTION
**Pull request by JosephThomasParker**
Addresses issue #1753

### CODE REVIEW

**Comment by ZedThree** [comment]
You're just missing the equivalent change in `gridfromfile.cxx`

File: src/mesh/mesh.cxx [hunk 1]
```diff
@@ -58,11 +58,14 @@ int Mesh::get(std::string &sval, const std::string &name) {
...
 
 /// Get an integer
-int Mesh::get(int &ival, const std::string &name) {
+int Mesh::get(int &ival, const std::string &name, int def) {
   TRACE("Mesh::get(ival, %s)", name.c_str());
 
-  if (source == nullptr or !source->get(this, ival, name))
+  if (source == nullptr or !source->get(this, ival, name, def)) {
+    // set ival to default in source==nullptr too:
+    ival = def;
```
Comment (JosephThomasParker) [inline src/mesh/mesh.cxx]: This is ugly: if `source == nullptr` then the default doesn't get set by `get`... so we set it silently here.  Without this, four of the integrated tests fail.

Comment (ZedThree) [reply src/mesh/mesh.cxx]: Could split it into two separate conditionals:

```cpp
if (source == nullptr) {
   output_warn << "No source; setting default"; // but better, obvs
   ival = def;
   return 1;
}
return source->get(this, ival, name, def);
```

**Comment by ZedThree** [comment]
Should we also add default-defaults for the other overloads that are missing them?

**Comment by JosephThomasParker** [comment]
Yes, if you're fine with the silent default-defaults.

**Comment by ZedThree** [comment]
The `Mesh::get` overloads for `string`, `BoutReal`, and `Vector2D/3D` could also do with defaults.

File: src/mesh/data/gridfromfile.cxx [hunk 2]
```diff
@@ -134,7 +134,7 @@ bool GridFile::get(Mesh *UNUSED(m), std::string &sval, const std::string &name)
  *   Boolean. True on success.
  * 
  */
-bool GridFile::get(Mesh *UNUSED(m), int &ival, const std::string &name) {
+bool GridFile::get(Mesh *UNUSED(m), int &ival, const std::string &name, int UNUSED(def)) {
```
Comment (ZedThree) [inline src/mesh/data/gridfromfile.cxx]: We should be able to use `def` to set the default here

File: src/mesh/mesh.cxx (hunk 1 above)
Comment (ZedThree) [reply src/mesh/mesh.cxx]: `GridDataSource::get` returns `true` on success, whereas `Mesh::get` returns false. So this just needed `return !source->get(...);`

**Comment by ZedThree** [comment]
I thought this was a nice simplification of the implementation of the `GridFromOptions::get` routines:

https://github.com/boutproject/BOUT-dev/blob/8bac5c5640cca5d6fc76a668f95b4d054927e079/src/mesh/data/gridfromoptions.cxx#L17-L21

We want to know if the source actually had the variable, or if we've ended up using the default. So save the result of `hasVar` in order to return it. Then just let `options` do the work of either getting us the value or the default.

This, rather unfortunately, leads to some rather convoluted bugs. Take this bit of code, for example:

https://github.com/boutproject/BOUT-dev/blob/e0310e96404f808e52444868807696fdf6e2c3e9/src/mesh/impls/bout/boutmesh.cxx#L130-L137

If the mesh source is a file, then if `MXG` isn't in the file, we read from the `options` instead, with a default of `2`. We also save the value of `2` in `options`, annotating that it was a default value.

Now, if the mesh source is an `Options` and `MXG` isn't in it, we instead record `0` in `options`, as this is the default-default. `Mesh::get` still returns `true` because it was a default, so we still go into if-body, where we get a new default of `2`, which we actually use. _However_, this doesn't overwrite the previous default of `0` we recorded in `options`. So when we write out `BOUT.settings` at the end of the simulation, it gets `MXG = 0`. This breaks the `test-command-args`, because it uses the `BOUT.settings` from one run as the input for another, and currently `MXG = 0` is not valid (without periodic X).

I think it needs to go back to being something like:

```cpp
 bool GridFromOptions::get(Mesh* UNUSED(m), int& ival, const std::string& name, int def) { 
   if (!hasVar(name)) { 
     output_warn.write("Variable '%s' not in mesh options. Setting to %d\n", name.c_str(), 
                       def); 
     ival = def; 
     return false; 
   } 
  
   ival = (*options)[name]; // No default!
   return true; 
 } 
```

But this is important to bear in mind if we replace `GridDataSource` completely with something `Options`-based.

**Pull request merged**